Sources/AsyncHTTPClient/HTTPDecompressor.swift

```diff
@@ -0,0 +1,151 @@
+    typealias OutboundIn = HTTPClientRequestPart
+    typealias OutboundOut = HTTPClientRequestPart
+
+    private enum State {
+        case empty
+        case compressed(CompressionAlgorithm, Int)
+    }
+
+    private let limit: HTTPClient.DecompressionLimit
+    private var state = State.empty
+    private var stream = z_stream()
+    private var inflated = 0
+
+    init(limit: HTTPClient.DecompressionLimit) {
+        self.limit = limit
+    }
+
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let request = self.unwrapOutboundIn(data)
+        switch request {
+        case .head(var head):
+            if head.headers.contains(name: "Accept-Encoding") {
+                context.write(data, promise: promise)
+            } else {
+                head.headers.replaceOrAdd(name: "Accept-Encoding", value: "deflate, gzip")
+                context.write(self.wrapOutboundOut(.head(head)), promise: promise)
+            }
+        default:
+            context.write(data, promise: promise)
+        }
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            let algorithm: CompressionAlgorithm?
+            let contentType = head.headers[canonicalForm: "Content-Encoding"].first?.lowercased()
+            if contentType == "gzip" {
+                algorithm = .gzip
+            } else if contentType == "deflate" {
+                algorithm = .deflate
+            } else {
+                algorithm = nil
+            }
+
+            let length = head.headers[canonicalForm: "Content-Length"].first.flatMap { Int($0) }
+
+            if let algorithm = algorithm, let length = length {
+                self.initializeDecoder(encoding: algorithm, length: length)
+            }
+
+            context.fireChannelRead(data)
+        case .body(var part):
+            switch self.state {
+            case .compressed(_, let originalLength):
+                self.stream.inflatePart(input: &part, allocator: context.channel.allocator) { output in
+                    self.inflated += output.readableBytes
+                    if self.limit.exceeded(compressed: originalLength, decompressed: self.inflated) {
+                        context.fireErrorCaught(HTTPClientError.decompressionLimit)
+                        return
+                    }
+                    context.fireChannelRead(self.wrapInboundOut(.body(output)))
+                }
+            default:
+                context.fireChannelRead(data)
+            }
+        case .end:
+            deflateEnd(&self.stream)
+            context.fireChannelRead(data)
+        }
+    }
+
+    private func initializeDecoder(encoding: CompressionAlgorithm, length: Int) {
+        self.state = .compressed(encoding, length)
+
+        self.stream.zalloc = nil
+        self.stream.zfree = nil
+        self.stream.opaque = nil
```

Comment:
Rather than explicitly nil these out, you could simply re-initialize the encoder entirely. It won't cost much more and you'll guarantee you're safe.

Reply:
I was following this:
https://github.com/apple/swift-nio-extras/blob/master/Sources/NIOHTTPCompression/HTTPResponseCompressor.swift#L193 that is following this, I think:
> The zalloc, zfree, and opaque fields in the strm structure must be initialized before calling deflateInit(). Here they are set to the zlib constant Z_NULL to request that zlib use the default memory allocation routines.

Reply:
see comment above. feels to me we should move the unsafe code into a separate library and  nio relevant extension into nio extras so its reusable beyond http client and the unsafe code concentrated in one place

Reply:
I am moderately nervous about the idea of NIO being a compression library. Is it worth considering whether we should split this out into a separate SSWG space altogether?

Reply:
exactly. sorry if my proposal ^^ was not clear enough, what i am suggesting is:
1. create a standalone compression library, sswg seems like a good home
2. have nio-extras module that helps with http compression, that utilizes that library 

seems like the most reusability. wdyt?

Reply:
Works for me, I think it's a good fit.

Reply:
Two questions:
1. what timeframe should we aim for? Can we postpone it until after `1.0.0` release?
2. should standalone compression library know about `ByteBuffer`? If not, what should we expose in the API?

Reply:
@tomerd @Lukasa ^^

Reply:
@artemredkin i suggest we start with moving all the compression related code to a separate internal/private module so we get the abstraction right. then we can evaluate what it means to spin off a separate library

re `ByteBuffer`, my gut feeling is a general purpose library should probably operate on native types. @Lukasa and @weissi probably probably also have insight on this

Reply:
The general purpose library should probably follow the pattern we use elsewhere: offer an implementation that works on `Collection where Element == UInt8` and that has a branch based on [`withContiguousStorageIfAvailable`](https://developer.apple.com/documentation/swift/sequence/3128824-withcontiguousstorageifavailable) that can do less shuffling.

That will allow `ByteBufferView` to be used here, which is sufficient for good performance with NIO.

Reply:
@Lukasa @weissi @tomerd we actually have an issue where we cannot have API for writing to a byte buffer, since we don't want to expose `Unsafe` API on the top level I propose to move all unsafe code to `nio-extras` (where we already have unsafe zlib support). That will allow us keep `async-http-client` safe and still use `ByteBuffer` in the compression/decompression APIs. PR is here (basically a copy of the current PR):
https://github.com/apple/swift-nio-extras/pull/56